// html5-player/libs/model/picture-quality.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.namespace = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

exports.default = function () {
  return {
    namespace: namespace,
    state: {
      list: null,
      //-1为自动
      currentQuality: -1
    },
    reducers: {
      dataReducer: _util.standardReducer,
      clear: function clear(state, _ref) {
        var payload = _ref.payload;

        return this.state;
      }
    },
    sagas: {
      dataSaga: /*#__PURE__*/_regenerator2.default.mark(function dataSaga(_ref2, _ref3) {
        var payload = _ref2.payload;
        var put = _ref3.put;
        return _regenerator2.default.wrap(function dataSaga$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return put({
                  type: 'dataReducer',
                  payload: payload
                });

              case 2:
              case 'end':
                return _context.stop();
            }
          }
        }, dataSaga, this);
      })
    }
  };
};

var _util = require('../utils/util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var namespace = exports.namespace = 'picture-quality';
