// html5-player/libs/model/time-slider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.namespace = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

exports.default = function () {
  return {
    namespace: namespace,
    state: {
      //最大值为1
      position: 0,
      buffer: 0,
      duration: 0
    },
    reducers: {
      timeReducer: _util.standardReducer,
      clear: function clear(state, _ref) {
        var payload = _ref.payload;

        return this.state;
      }
    },
    sagas: {
      timeSaga: /*#__PURE__*/_regenerator2.default.mark(function timeSaga(_ref2, _ref3) {
        var _ref2$payload = _ref2.payload,
            currentTime = _ref2$payload.currentTime,
            _ref2$payload$duratio = _ref2$payload.duration,
            duration = _ref2$payload$duratio === undefined ? 0 : _ref2$payload$duratio,
            buffer = _ref2$payload.buffer;
        var put = _ref3.put;
        var percent;
        return _regenerator2.default.wrap(function timeSaga$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                percent = currentTime / duration;
                // console.log(percent, currentTime, duration);

                if (percent > 1) {
                  percent = 1;
                }
                if (isNaN(percent)) {
                  percent = 0;
                }
                if (isNaN(duration)) {
                  duration = 0;
                }
                _context.next = 6;
                return put({
                  type: 'timeReducer',
                  payload: {
                    percent: percent,
                    buffer: buffer / duration,
                    duration: duration
                  }
                });

              case 6:
              case 'end':
                return _context.stop();
            }
          }
        }, timeSaga, this);
      })
    }
  };
};

var _util = require('../utils/util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var namespace = exports.namespace = 'time-slider';
