// html5-player/libs/model/volume.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.namespace = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

exports.default = function () {
  return {
    namespace: namespace,
    //100为自定义的总音量，原生的最大为1
    //先设置声音为20，太大可能一开始就造成声音过多，用户体验不好。
    state: 20,
    reducers: {
      dataReducer: function dataReducer(state, _ref) {
        var payload = _ref.payload;

        return payload;
      },
      clear: function clear(state, _ref2) {
        var payload = _ref2.payload;

        return this.state;
      }
    },
    sagas: {
      dataSaga: /*#__PURE__*/_regenerator2.default.mark(function dataSaga(_ref3, _ref4) {
        var payload = _ref3.payload;
        var put = _ref4.put;
        return _regenerator2.default.wrap(function dataSaga$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (payload && isNaN(payload)) {
                  //兼容volume不为数字的情况
                  payload = 0;
                }
                _context.next = 3;
                return put({
                  type: 'dataReducer',
                  payload: payload
                });

              case 3:
              case 'end':
                return _context.stop();
            }
          }
        }, dataSaga, this);
      })
    }
  };
};

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// import { standardReducer } from '../utils/util';

var namespace = exports.namespace = 'volume';
