// html5-player/libs/view/history/time-slider/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _dec2, _class, _class2, _temp2; //外部依赖包

//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactRedux = require('react-redux');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _clear = require('../../decorator/clear');

var _clear2 = _interopRequireDefault(_clear);

var _slider = require('../../components/slider');

var _slider2 = _interopRequireDefault(_slider);

var _timeTooltip = require('./time-tooltip');

var _timeTooltip2 = _interopRequireDefault(_timeTooltip);

var _selection = require('./selection');

var _selection2 = _interopRequireDefault(_selection);

var _video = require('../../../model/video');

var _timeSlider = require('../../../model/time-slider');

var _selection3 = require('../../../model/selection');

var _errorMessage = require('../../../model/error-message');

var _end = require('../../../model/end');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 播放器加载状态的组件
 */
var TimeSlider = (_dec = (0, _reactRedux.connect)(function (state) {
  return {
    sliderModel: state[_timeSlider.namespace],
    selection: state[_selection3.namespace],
    isError: !!state[_errorMessage.namespace].message,
    end: state[_end.namespace]
  };
}), _dec2 = (0, _clear2.default)([_timeSlider.namespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(TimeSlider, _React$Component);

  function TimeSlider() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, TimeSlider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = TimeSlider.__proto__ || (0, _getPrototypeOf2.default)(TimeSlider)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _this.dispatch = _this.props.dispatch, _this.onSliderChange = function (percent) {
      var _this$props = _this.props,
          activeItem = _this$props.activeItem,
          fragments = _this$props.fragments,
          storage = _this$props.storage,
          sliderModel = _this$props.sliderModel,
          selection = _this$props.selection;

      if (selection && selection.seekingDisabled) {
        return false;
      }
      var _sliderModel$duration = sliderModel.duration,
          currentVideoDuration = _sliderModel$duration === undefined ? 0 : _sliderModel$duration;

      var duration = _this.duration;
      if (currentVideoDuration > duration) {
        //只有一个视频的时候，会存在误差
        duration = currentVideoDuration;
      }
      var currentTime = percent * duration;
      //当前播放的currentTime，需要转换
      var currentVideoTime = currentTime;
      _this.selectedItem = activeItem;
      fragments.forEach(function (v, k) {
        if (v.begin < currentTime && v.end > currentTime) {
          //切换视频会用到
          _this.selectedItem = k;
          return;
        }
      });
      fragments.forEach(function (v, k) {
        //这个循环需要等待selected赋值后处理，即上面的循环处理
        var gaps = v.end - v.begin;
        if (_this.selectedItem > k) {
          currentVideoTime -= gaps;
        }
      });
      //存储storage
      storage.lastPercent = percent;
      _this.setState({
        percent: percent
      });
      //begin----处理当前video的进度
      var broken = !fragments[_this.selectedItem].file;
      if (broken) {
        //拖动到broken或者点击到broken处（即无视频处）
        currentVideoTime = 0;
      }
      _this.currentVideoTime = currentVideoTime;

      _this.dispatch({
        type: _video.namespace + '/seeking',
        payload: {
          percent: _this.currentVideoTime / currentVideoDuration
        }
      });
      //end----处理当前video的进度
    }, _this.onMouseDown = function (e) {
      var selection = _this.props.selection;

      if (selection && selection.seekingDisabled) {
        return false;
      }
      _this.hasbeenClick = true;
      _this.dispatch({
        type: _video.namespace + '/seekingState',
        payload: true
      });
      _this.setState({
        isSliding: true
      });
    }, _this.onMouseUp = function (e) {
      var selection = _this.props.selection;

      if (selection && selection.seekingDisabled) {
        return false;
      }
      _this.setState({
        isSliding: false
      });
      var _this$props2 = _this.props,
          storage = _this$props2.storage,
          setActiveItem = _this$props2.setActiveItem,
          activeItem = _this$props2.activeItem;
      //video/event/index.js中loadeddata事件中使用。

      window.historyVideoCurrentTime = _this.currentVideoTime;
      _this.dispatch({
        type: _video.namespace + '/seekingState',
        payload: false
      });
      if (activeItem !== _this.selectedItem) {
        //拖动或者点击slider，选中的item不一样，需要切换item进行播放
        setActiveItem(_this.selectedItem);
        //视频切换需要设置defaultCurrentTime，setActiveItem会先重置defaultCurrentTime=0。
        storage.defaultCurrentTime = _this.currentVideoTime;
      }
      if (_this.selectedItem !== 0) {
        _this.setState({
          percent: 0
        });
      }
    }, _this.onLeftSelectionBlur = function (percent) {
      _this.dispatch({
        type: _video.namespace + '/selection',
        payload: {
          type: 'left-blur',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _this.onLeftSelectionChange = function (percent) {
      _this.dispatch({
        type: _video.namespace + '/selection',
        payload: {
          type: 'left-change',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _this.onRightSelectionChange = function (percent) {
      _this.dispatch({
        type: _video.namespace + '/selection',
        payload: {
          type: 'right-change',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _this.onRightSelectionBlur = function (percent) {
      _this.dispatch({
        type: _video.namespace + '/selection',
        payload: {
          type: 'right-blur',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(TimeSlider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        timeSliderDOM: this.timeSliderDOM
      };
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.timeSliderDOM = _reactDom2.default.findDOMNode(this.refs.slider);
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      //为了对外提供api，historyCurrentTime
      this.dispatch({
        type: _video.namespace + '/setHistoryCurrentTime',
        payload: {
          historyCurrentTime: this.percent * this.duration
        }
      });
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.historyVideoCurrentTime = 0;
    }
  }, {
    key: 'renderSelection',

    /**
     * 在有fragment的前提下，渲染可以选择播放的操作按钮
     */
    value: function renderSelection() {
      var selection = this.props.selection;

      if (!selection) {
        return false;
      }
      var begin = selection.begin,
          end = selection.end;

      var duration = this.duration;
      if (end > duration) {
        end = duration;
      }
      var leftPercent = begin / duration;
      var rightPercent = end / duration;
      return _react2.default.createElement(
        'span',
        { className: 'html5-player-selection-container' },
        _react2.default.createElement(
          _selection2.default,
          {
            className: 'html5-player-selection-left',
            percent: leftPercent,
            onBlur: this.onLeftSelectionBlur,
            onChange: this.onLeftSelectionChange
          },
          this.context.leftSelectionComponent
        ),
        _react2.default.createElement(
          _selection2.default,
          {
            percent: rightPercent,
            className: 'html5-player-selection-right',
            onChange: this.onRightSelectionChange,
            onBlur: this.onRightSelectionBlur
          },
          this.context.rightSelectionComponent
        )
      );
    }
  }, {
    key: 'getPercent',
    value: function getPercent() {
      var _props = this.props,
          sliderModel = _props.sliderModel,
          fragments = _props.fragments,
          activeItem = _props.activeItem,
          storage = _props.storage,
          end = _props.end;
      //currentVideoDuration当前播放视频的duration

      var _sliderModel$percent = sliderModel.percent,
          currentVideoPercent = _sliderModel$percent === undefined ? 0 : _sliderModel$percent,
          _sliderModel$duration2 = sliderModel.duration,
          currentVideoDuration = _sliderModel$duration2 === undefined ? 0 : _sliderModel$duration2;

      var duration = this.duration;
      if ((currentVideoPercent === 0 || currentVideoDuration === 0) && activeItem > 0) {
        //切换视频初始化时(获取数据)，返回上一个percent
        //这样才能位置进度不变，要不会变成进度为0，然后视频加载成功后跳跃到播放位置，这样就很奇怪
        return storage.lastPercent || 0;
      }
      if (
      //slider拖动中
      this.state.isSliding ||
      //下面的条件判断，处理切换到第一个视频初始化时，进度为0的问题
      activeItem === 0 && currentVideoPercent === 0) {
        return this.state.percent || 0;
      }
      // console.log(currentVideoPercent);
      var percent = 0;
      //当前播放的视频的当前播放时长
      var currentVideoTime = currentVideoPercent * currentVideoDuration;
      if (currentVideoTime === currentVideoDuration && activeItem < fragments.length && !end) {
        //处理当前视频播放结束切换下一个视频时，currentVideoTime还等于上一个的duration问题。
        currentVideoTime = 0;
      }
      if (window.historyVideoCurrentTime) {
        window.historyVideoCurrentTime = 0;
      }
      // console.log(window.historyVideoCurrentTime);
      var currentTime = 0;
      currentTime += currentVideoTime;
      fragments.forEach(function (v, k) {
        if (activeItem > k) {
          var gap = v.end - v.begin;
          currentTime += gap;
        }
      });
      percent = currentTime / duration;
      storage.lastPercent = percent;
      if (percent > 1) {
        percent = 1;
      }
      return percent;
    }
    //是否是最后一个可以播放的视频

  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          fragments = _props2.fragments,
          beginDateTime = _props2.beginDateTime,
          selection = _props2.selection,
          isError = _props2.isError,
          end = _props2.end;

      var duration = this.duration;
      // this.setStorage();
      var percent = this.getPercent();
      this.percent = percent;
      return _react2.default.createElement(
        _slider2.default,
        {
          ref: 'slider',
          onChange: this.onSliderChange,
          percent: percent,
          padding: '0 10px',
          className: (0, _classnames2.default)('html5-player-time-slider', 'html5-player-track-history-slider', {
            'html5-player-hide': isError || this.isLastVideo && end
          }),
          onMouseDown: this.onMouseDown,
          onMouseUp: this.onMouseUp,
          onlyShowCircleOnEnter: true
        },
        selection && this.renderSelection(),
        fragments && fragments.map(function (v, k) {
          var gap = v.end - v.begin;
          if (fragments.length === k + 1) {
            //最后一个加多一个1，这样才会覆盖整个slider
            gap += 1;
          }
          var width = gap / duration * 100 + '%';
          var left = v.begin / duration * 100 + '%';
          var style = {
            left: left,
            width: width
          };
          return _react2.default.createElement('div', {
            key: k,
            className: (0, _classnames2.default)({
              'html5-player-broken': !v.file,
              'html5-player-hasvideo': v.file
            }),
            style: style
          });
        }),
        _react2.default.createElement(_timeTooltip2.default, { duration: duration, beginDateTime: beginDateTime })
      );
    }
  }, {
    key: 'duration',
    get: function get() {
      var _props3 = this.props,
          duration = _props3.duration,
          sliderModel = _props3.sliderModel;
      var _sliderModel$duration3 = sliderModel.duration,
          currentVideoDuration = _sliderModel$duration3 === undefined ? 0 : _sliderModel$duration3;

      if (currentVideoDuration > duration) {
        //只有一个视频的时候，会存在误差
        duration = currentVideoDuration;
      }
      return duration;
    }
  }, {
    key: 'isLastVideo',
    get: function get() {
      var _props4 = this.props,
          fragments = _props4.fragments,
          activeItem = _props4.activeItem;

      activeItem += 1;
      if (fragments[activeItem] && !fragments[activeItem].file && fragments.length - 1 === activeItem) {
        //判断下一个是否有视频（最后一个视频）
        return true;
      }
      if (fragments && fragments[0] && activeItem < fragments.length) {
        return false;
      }
      return true;
    }
  }]);
  return TimeSlider;
}(_react2.default.Component), _class2.propTypes = {}, _class2.displayName = 'TimeSlider', _class2.contextTypes = {
  leftSelectionComponent: _propTypes2.default.element,
  rightSelectionComponent: _propTypes2.default.element
}, _class2.childContextTypes = {
  timeSliderDOM: _propTypes2.default.object
}, _temp2)) || _class) || _class);
exports.default = TimeSlider;
