// html5-player/libs/view/history/time-slider/time-tooltip.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp2; //外部依赖包

//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _tooltip = require('../../components/tooltip');

var _tooltip2 = _interopRequireDefault(_tooltip);

var _util = require('../../../utils/util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TimeTooltip = (_temp2 = _class = function (_React$Component) {
  (0, _inherits3.default)(TimeTooltip, _React$Component);

  function TimeTooltip() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, TimeTooltip);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = TimeTooltip.__proto__ || (0, _getPrototypeOf2.default)(TimeTooltip)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      percent: 0.5
    }, _this.dispatch = _this.props.dispatch, _this.onChange = function (percent) {
      _this.setState({
        percent: percent
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }
  //这里的配置参考jw-player的api


  (0, _createClass3.default)(TimeTooltip, [{
    key: 'renderContent',
    value: function renderContent() {
      var _props = this.props,
          duration = _props.duration,
          beginDateTime = _props.beginDateTime;
      var percent = this.state.percent;

      if (duration === 0) {
        return false;
      }
      var formatTime = beginDateTime && (0, _util.dateFormat)((beginDateTime + percent * duration) * 1000, 'YYYY-MM-DD HH:mm:ss');
      return _react2.default.createElement(
        'div',
        { className: 'html5-player-time-tooltip-content' },
        formatTime
      );
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        _tooltip2.default,
        {
          className: 'html5-player-time-tooltip',
          content: this.renderContent(),
          type: 'move',
          onChange: this.onChange,
          percent: this.state.percent,
          isAnimateActive: false
        },
        _react2.default.createElement('div', { className: 'html5-player-for-tooltip' })
      );
    }
  }]);
  return TimeTooltip;
}(_react2.default.Component), _class.propTypes = {}, _class.displayName = 'TimeTooltip', _temp2);
exports.default = TimeTooltip;
