// html5-player/es/model/error-message.js
import _regeneratorRuntime from 'babel-runtime/regenerator';
export var namespace = 'error-message';
export default function () {
  return {
    namespace: namespace,
    state: {
      message: null
    },
    reducers: {
      errorMessageReducer: function errorMessageReducer(state, _ref) {
        var payload = _ref.payload;

        return payload;
      },
      clear: function clear(state, _ref2) {
        var payload = _ref2.payload;

        return this.state;
      }
    },
    sagas: {
      errorMessageSaga: /*#__PURE__*/_regeneratorRuntime.mark(function errorMessageSaga(_ref3, _ref4) {
        var payload = _ref3.payload;
        var put = _ref4.put;
        return _regeneratorRuntime.wrap(function errorMessageSaga$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return put({
                  type: 'errorMessageReducer',
                  payload: payload
                });

              case 2:
              case 'end':
                return _context.stop();
            }
          }
        }, errorMessageSaga, this);
      })
    }
  };
}
