// html5-player/es/model/volume.js
import _regeneratorRuntime from 'babel-runtime/regenerator';
// import { standardReducer } from '../utils/util';

export var namespace = 'volume';
export default function () {
  return {
    namespace: namespace,
    //100为自定义的总音量，原生的最大为1
    //先设置声音为20，太大可能一开始就造成声音过多，用户体验不好。
    state: 20,
    reducers: {
      dataReducer: function dataReducer(state, _ref) {
        var payload = _ref.payload;

        return payload;
      },
      clear: function clear(state, _ref2) {
        var payload = _ref2.payload;

        return this.state;
      }
    },
    sagas: {
      dataSaga: /*#__PURE__*/_regeneratorRuntime.mark(function dataSaga(_ref3, _ref4) {
        var payload = _ref3.payload;
        var put = _ref4.put;
        return _regeneratorRuntime.wrap(function dataSaga$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (payload && isNaN(payload)) {
                  //兼容volume不为数字的情况
                  payload = 0;
                }
                _context.next = 3;
                return put({
                  type: 'dataReducer',
                  payload: payload
                });

              case 3:
              case 'end':
                return _context.stop();
            }
          }
        }, dataSaga, this);
      })
    }
  };
}
