// html5-player/libs/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp2; //外部依赖包

//内部依赖包

//icon的js


exports.default = player;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

require('./version');

var _sagaModelProvider = require('./libs/provider/saga-model-provider');

var _sagaModelProvider2 = _interopRequireDefault(_sagaModelProvider);

var _modelList = require('./model-list');

var _modelList2 = _interopRequireDefault(_modelList);

var _view = require('./view');

var _view2 = _interopRequireDefault(_view);

require('./assets/icon/iconfont');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ModelRegister = (_temp2 = _class = function (_React$Component) {
  (0, _inherits3.default)(ModelRegister, _React$Component);

  function ModelRegister() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ModelRegister);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ModelRegister.__proto__ || (0, _getPrototypeOf2.default)(ModelRegister)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'ModelRegister', _this.state = {}, _this.registerModel = function (register) {
      var allModelPromise = _modelList2.default.map(function (modelId) {
        var model = require('./model/' + modelId + '.js').default;
        return model;
      });
      return _promise2.default.all(allModelPromise).then(function (models) {
        models.forEach(function (m) {
          register(m);
        });
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ModelRegister, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      this.registerModel(this.context.sagaStore.register).then(function () {
        _this2.setState({
          canBeRendered: true
        });
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var children = this.props.children;

      if (this.state.canBeRendered) {
        return _react2.default.createElement(
          'span',
          null,
          children
        );
      } else {
        return false;
      }
    }
  }]);
  return ModelRegister;
}(_react2.default.Component), _class.contextTypes = {
  sagaStore: _propTypes2.default.object
}, _temp2);
function player(props) {
  return _react2.default.createElement(
    _sagaModelProvider2.default,
    {
      production: process.env.NODE_ENV === 'production',
      plugins: [{
        onError: function onError(error, dispatch) {
          console.error(error);
        }
      }]
    },
    _react2.default.createElement(
      ModelRegister,
      null,
      _react2.default.createElement(_view2.default, props)
    )
  );
}
