// html5-player/libs/model/end.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.namespace = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

exports.default = function () {
  return {
    namespace: namespace,
    state: false,
    reducers: {
      endStateReducer: function endStateReducer(state, _ref) {
        var payload = _ref.payload;

        return payload;
      },
      clear: function clear(state, _ref2) {
        var payload = _ref2.payload;

        return this.state;
      }
    },
    sagas: {
      endStateSaga: /*#__PURE__*/_regenerator2.default.mark(function endStateSaga(_ref3, _ref4) {
        var payload = _ref3.payload;
        var put = _ref4.put;
        return _regenerator2.default.wrap(function endStateSaga$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return put({
                  type: 'endStateReducer',
                  payload: payload
                });

              case 2:
              case 'end':
                return _context.stop();
            }
          }
        }, endStateSaga, this);
      })
    }
  };
};

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var namespace = exports.namespace = 'end';
