// html5-player/es/model/play-pause.js
import _regeneratorRuntime from 'babel-runtime/regenerator';
export var namespace = 'play-pause';
export default function () {
  return {
    namespace: namespace,
    //是否播放
    state: false,
    reducers: {
      playPauseReducer: function playPauseReducer(state, _ref) {
        var payload = _ref.payload;

        return payload;
      },
      clear: function clear(state, _ref2) {
        var payload = _ref2.payload;

        return this.state;
      }
    },
    sagas: {
      playPauseSaga: /*#__PURE__*/_regeneratorRuntime.mark(function playPauseSaga(_ref3, _ref4) {
        var payload = _ref3.payload;
        var put = _ref4.put;
        return _regeneratorRuntime.wrap(function playPauseSaga$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return put({
                  type: 'playPauseReducer',
                  payload: payload
                });

              case 2:
              case 'end':
                return _context.stop();
            }
          }
        }, playPauseSaga, this);
      })
    }
  };
}
