// html5-player/es/model/rotate.js
import _regeneratorRuntime from 'babel-runtime/regenerator';
export var namespace = 'rotate';
export default function () {
  return {
    namespace: namespace,
    state: false,
    reducers: {
      dataReducer: function dataReducer(state, _ref) {
        var payload = _ref.payload;

        return payload;
      },
      clear: function clear(state, _ref2) {
        var payload = _ref2.payload;

        return this.state;
      }
    },
    sagas: {
      dataSaga: /*#__PURE__*/_regeneratorRuntime.mark(function dataSaga(_ref3, _ref4) {
        var payload = _ref3.payload;
        var put = _ref4.put;
        return _regeneratorRuntime.wrap(function dataSaga$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return put({
                  type: 'dataReducer',
                  payload: payload
                });

              case 2:
              case 'end':
                return _context.stop();
            }
          }
        }, dataSaga, this);
      })
    }
  };
}
