// html5-player/es/model/time.js
import _regeneratorRuntime from 'babel-runtime/regenerator';
import { standardReducer, hms } from '../utils/util';

export var namespace = 'time';
export default function () {
  return {
    namespace: namespace,
    state: {
      currentTime: 0,
      elapse: '00:00',
      duration: '00:00',
      //原始video的duration，为了让其他地方使用，存一份
      secondDuration: 0
    },
    reducers: {
      timeReducer: standardReducer,
      clear: function clear(state, _ref) {
        var payload = _ref.payload;

        return this.state;
      }
    },
    sagas: {
      timeSaga: /*#__PURE__*/_regeneratorRuntime.mark(function timeSaga(_ref2, _ref3) {
        var _ref2$payload = _ref2.payload,
            currentTime = _ref2$payload.currentTime,
            _ref2$payload$duratio = _ref2$payload.duration,
            duration = _ref2$payload$duratio === undefined ? 0 : _ref2$payload$duratio;
        var put = _ref3.put;
        var elapse, timeDuration;
        return _regeneratorRuntime.wrap(function timeSaga$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                //currentTime和duration会有误差，最大的currentTime基本都比duration小，+0.25是为了减少time中的误差。
                elapse = hms(Math.min(currentTime + 0.25, duration));
                timeDuration = hms(duration);
                _context.next = 4;
                return put({
                  type: 'timeReducer',
                  payload: {
                    elapse: elapse,
                    duration: timeDuration,
                    secondDuration: duration,
                    currentTime: currentTime
                  }
                });

              case 4:
              case 'end':
                return _context.stop();
            }
          }
        }, timeSaga, this);
      })
    }
  };
}
