// html5-player/libs/view/controlbar/setting/list.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp2; //外部依赖包

//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _video = require('../../../model/video');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var List = (_temp2 = _class = function (_React$Component) {
  (0, _inherits3.default)(List, _React$Component);

  function List() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, List);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = List.__proto__ || (0, _getPrototypeOf2.default)(List)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'List', _this.dispatch = _this.props.dispatch, _this.onSelectEvent = function (value, callback) {
      return function (e) {
        var onSelect = _this.props.onSelect;

        onSelect && onSelect(value, e);
        callback && callback();
        if (!_this.context.playerDOM.paused) {
          //播放才处理，选择的时候因为鼠标位置已经离开了controlbar
          //所有选择完成需要，触发隐藏controlbar
          _this.dispatch({
            type: _video.namespace + '/controlbar',
            payload: false,
            delayTime: _this.context.controlbarHideTime,
            onControlbarEnter: false
          });
        }
      };
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(List, [{
    key: 'getLocale',
    value: function getLocale() {
      return this.context.localization;
    }
  }, {
    key: 'renderBack',
    value: function renderBack(title) {
      var onBackEvent = this.props.onBackEvent;

      if (!onBackEvent) {
        return false;
      } else {
        return _react2.default.createElement(
          'li',
          { className: 'html5-player-list-title', onClick: onBackEvent },
          _react2.default.createElement(
            'svg',
            {
              className: 'html5-player-icon html5-player-left-icon',
              'aria-hidden': 'true'
            },
            _react2.default.createElement('use', { xlinkHref: '#icon-left' })
          ),
          title
        );
      }
    }
  }]);
  return List;
}(_react2.default.Component), _class.contextTypes = {
  localization: _propTypes2.default.object,
  playerDOM: _propTypes2.default.object,
  controlbarHideTime: _propTypes2.default.number
}, _temp2);
exports.default = List;
