// html5-player/libs/view/controlbar/setting/picture-quality-list.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class; //外部依赖包

//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _list = require('./list');

var _list2 = _interopRequireDefault(_list);

var _video = require('../../../model/video');

var _pictureQuality = require('../../../model/picture-quality');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PictureQualityList = (_dec = (0, _reactRedux.connect)(function (state) {
  return {
    qualityList: state[_pictureQuality.namespace].list,
    currentQuality: state[_pictureQuality.namespace].currentQuality
  };
}), _dec(_class = function (_List) {
  (0, _inherits3.default)(PictureQualityList, _List);

  function PictureQualityList() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, PictureQualityList);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = PictureQualityList.__proto__ || (0, _getPrototypeOf2.default)(PictureQualityList)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'PictureQualityList', _this.onSelect = function (value) {
      return _this.onSelectEvent(value, function () {
        _this.dispatch({
          type: _video.namespace + '/switchPictureQuality',
          payload: value
        });
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(PictureQualityList, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          qualityList = _props.qualityList,
          currentQuality = _props.currentQuality;

      var locale = this.getLocale();
      return _react2.default.createElement(
        'ul',
        { className: 'html5-player-list-container' },
        this.renderBack(locale.pictureQuality),
        qualityList && qualityList.map(function (v) {
          var className = (0, _classnames2.default)({
            'html5-player-list-selected': currentQuality === v.value
          });
          return _react2.default.createElement(
            'li',
            {
              onClick: _this2.onSelect(v.value),
              key: v.value,
              className: className
            },
            v.label
          );
        }),
        _react2.default.createElement(
          'li',
          {
            onClick: this.onSelect(-1),
            key: -1,
            className: (0, _classnames2.default)({
              'html5-player-list-selected': currentQuality === -1
            })
          },
          locale.autoQuality
        )
      );
    }
  }]);
  return PictureQualityList;
}(_list2.default)) || _class);
exports.default = PictureQualityList;
