// html5-player/es/libs/provider/saga-model.js
import _toConsumableArray from 'babel-runtime/helpers/toConsumableArray';
import _extends from 'babel-runtime/helpers/extends';
import { SagaModel } from 'redux-saga-model';

/**
 * 获取saga model实例化对象
 * @param { object } history 同react-router history
 * @param { object } reducers 为经过combineReducer的reducers
 * @param { any } preloadedState 请参考redux的createStore第二参数
 * @param { array } models 请参考redux-saga-model说明
 * @param { array } plugins 请参考redux-saga-model说明
 */
export default function getSagaModel(reducers, preloadedState) {
  var models = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var middlewares = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];
  var plugins = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : [];
  var openReduxDevTools = arguments[5];

  var initialState = preloadedState;
  var initialReducer = _extends({}, reducers);
  var initialMiddleware = [].concat(_toConsumableArray(middlewares));
  var initialModels = models;
  var sagaModel = new SagaModel({
    initialState: initialState,
    initialReducer: initialReducer,
    initialMiddleware: initialMiddleware,
    initialModels: initialModels,
    history: history,
    openReduxDevTools: openReduxDevTools
  });
  plugins.forEach(sagaModel.use.bind(sagaModel));
  return sagaModel;
}
