// html5-player/es/view/history/time-slider/index.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _dec, _dec2, _class, _class2, _temp2;

//外部依赖包
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import classnames from 'classnames';
//内部依赖包
import clearDecorator from '../../decorator/clear';
import Slider from '../../components/slider';
import TimeTooltip from './time-tooltip';
import Selection from './selection';
import { namespace as videoNamespace } from '../../../model/video';
import { namespace as timeSliderNamespace } from '../../../model/time-slider';
import { namespace as selectionNamespace } from '../../../model/selection';
import { namespace as errorNamespace } from '../../../model/error-message';
import { namespace as endNamespace } from '../../../model/end';

/**
 * 播放器加载状态的组件
 */
var TimeSlider = (_dec = connect(function (state) {
  return {
    sliderModel: state[timeSliderNamespace],
    selection: state[selectionNamespace],
    isError: !!state[errorNamespace].message,
    end: state[endNamespace]
  };
}), _dec2 = clearDecorator([timeSliderNamespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  _inherits(TimeSlider, _React$Component);

  function TimeSlider() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, TimeSlider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = TimeSlider.__proto__ || _Object$getPrototypeOf(TimeSlider)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _this.dispatch = _this.props.dispatch, _this.onSliderChange = function (percent) {
      var _this$props = _this.props,
          activeItem = _this$props.activeItem,
          fragments = _this$props.fragments,
          storage = _this$props.storage,
          sliderModel = _this$props.sliderModel,
          selection = _this$props.selection;

      if (selection && selection.seekingDisabled) {
        return false;
      }
      var _sliderModel$duration = sliderModel.duration,
          currentVideoDuration = _sliderModel$duration === undefined ? 0 : _sliderModel$duration;

      var duration = _this.duration;
      if (currentVideoDuration > duration) {
        //只有一个视频的时候，会存在误差
        duration = currentVideoDuration;
      }
      var currentTime = percent * duration;
      //当前播放的currentTime，需要转换
      var currentVideoTime = currentTime;
      _this.selectedItem = activeItem;
      fragments.forEach(function (v, k) {
        if (v.begin < currentTime && v.end > currentTime) {
          //切换视频会用到
          _this.selectedItem = k;
          return;
        }
      });
      fragments.forEach(function (v, k) {
        //这个循环需要等待selected赋值后处理，即上面的循环处理
        var gaps = v.end - v.begin;
        if (_this.selectedItem > k) {
          currentVideoTime -= gaps;
        }
      });
      //存储storage
      storage.lastPercent = percent;
      _this.setState({
        percent: percent
      });
      //begin----处理当前video的进度
      var broken = !fragments[_this.selectedItem].file;
      if (broken) {
        //拖动到broken或者点击到broken处（即无视频处）
        currentVideoTime = 0;
      }
      _this.currentVideoTime = currentVideoTime;

      _this.dispatch({
        type: videoNamespace + '/seeking',
        payload: {
          percent: _this.currentVideoTime / currentVideoDuration
        }
      });
      //end----处理当前video的进度
    }, _this.onMouseDown = function (e) {
      var selection = _this.props.selection;

      if (selection && selection.seekingDisabled) {
        return false;
      }
      _this.hasbeenClick = true;
      _this.dispatch({
        type: videoNamespace + '/seekingState',
        payload: true
      });
      _this.setState({
        isSliding: true
      });
    }, _this.onMouseUp = function (e) {
      var selection = _this.props.selection;

      if (selection && selection.seekingDisabled) {
        return false;
      }
      _this.setState({
        isSliding: false
      });
      var _this$props2 = _this.props,
          storage = _this$props2.storage,
          setActiveItem = _this$props2.setActiveItem,
          activeItem = _this$props2.activeItem;
      //video/event/index.js中loadeddata事件中使用。

      window.historyVideoCurrentTime = _this.currentVideoTime;
      _this.dispatch({
        type: videoNamespace + '/seekingState',
        payload: false
      });
      if (activeItem !== _this.selectedItem) {
        //拖动或者点击slider，选中的item不一样，需要切换item进行播放
        setActiveItem(_this.selectedItem);
        //视频切换需要设置defaultCurrentTime，setActiveItem会先重置defaultCurrentTime=0。
        storage.defaultCurrentTime = _this.currentVideoTime;
      }
      if (_this.selectedItem !== 0) {
        _this.setState({
          percent: 0
        });
      }
    }, _this.onLeftSelectionBlur = function (percent) {
      _this.dispatch({
        type: videoNamespace + '/selection',
        payload: {
          type: 'left-blur',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _this.onLeftSelectionChange = function (percent) {
      _this.dispatch({
        type: videoNamespace + '/selection',
        payload: {
          type: 'left-change',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _this.onRightSelectionChange = function (percent) {
      _this.dispatch({
        type: videoNamespace + '/selection',
        payload: {
          type: 'right-change',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _this.onRightSelectionBlur = function (percent) {
      _this.dispatch({
        type: videoNamespace + '/selection',
        payload: {
          type: 'right-blur',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(TimeSlider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        timeSliderDOM: this.timeSliderDOM
      };
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.timeSliderDOM = ReactDOM.findDOMNode(this.refs.slider);
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      //为了对外提供api，historyCurrentTime
      this.dispatch({
        type: videoNamespace + '/setHistoryCurrentTime',
        payload: {
          historyCurrentTime: this.percent * this.duration
        }
      });
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.historyVideoCurrentTime = 0;
    }
  }, {
    key: 'renderSelection',

    /**
     * 在有fragment的前提下，渲染可以选择播放的操作按钮
     */
    value: function renderSelection() {
      var selection = this.props.selection;

      if (!selection) {
        return false;
      }
      var begin = selection.begin,
          end = selection.end;

      var duration = this.duration;
      if (end > duration) {
        end = duration;
      }
      var leftPercent = begin / duration;
      var rightPercent = end / duration;
      return React.createElement(
        'span',
        { className: 'html5-player-selection-container' },
        React.createElement(
          Selection,
          {
            className: 'html5-player-selection-left',
            percent: leftPercent,
            onBlur: this.onLeftSelectionBlur,
            onChange: this.onLeftSelectionChange
          },
          this.context.leftSelectionComponent
        ),
        React.createElement(
          Selection,
          {
            percent: rightPercent,
            className: 'html5-player-selection-right',
            onChange: this.onRightSelectionChange,
            onBlur: this.onRightSelectionBlur
          },
          this.context.rightSelectionComponent
        )
      );
    }
  }, {
    key: 'getPercent',
    value: function getPercent() {
      var _props = this.props,
          sliderModel = _props.sliderModel,
          fragments = _props.fragments,
          activeItem = _props.activeItem,
          storage = _props.storage,
          end = _props.end;
      //currentVideoDuration当前播放视频的duration

      var _sliderModel$percent = sliderModel.percent,
          currentVideoPercent = _sliderModel$percent === undefined ? 0 : _sliderModel$percent,
          _sliderModel$duration2 = sliderModel.duration,
          currentVideoDuration = _sliderModel$duration2 === undefined ? 0 : _sliderModel$duration2;

      var duration = this.duration;
      if ((currentVideoPercent === 0 || currentVideoDuration === 0) && activeItem > 0) {
        //切换视频初始化时(获取数据)，返回上一个percent
        //这样才能位置进度不变，要不会变成进度为0，然后视频加载成功后跳跃到播放位置，这样就很奇怪
        return storage.lastPercent || 0;
      }
      if (
      //slider拖动中
      this.state.isSliding ||
      //下面的条件判断，处理切换到第一个视频初始化时，进度为0的问题
      activeItem === 0 && currentVideoPercent === 0) {
        return this.state.percent || 0;
      }
      // console.log(currentVideoPercent);
      var percent = 0;
      //当前播放的视频的当前播放时长
      var currentVideoTime = currentVideoPercent * currentVideoDuration;
      if (currentVideoTime === currentVideoDuration && activeItem < fragments.length && !end) {
        //处理当前视频播放结束切换下一个视频时，currentVideoTime还等于上一个的duration问题。
        currentVideoTime = 0;
      }
      if (window.historyVideoCurrentTime) {
        window.historyVideoCurrentTime = 0;
      }
      // console.log(window.historyVideoCurrentTime);
      var currentTime = 0;
      currentTime += currentVideoTime;
      fragments.forEach(function (v, k) {
        if (activeItem > k) {
          var gap = v.end - v.begin;
          currentTime += gap;
        }
      });
      percent = currentTime / duration;
      storage.lastPercent = percent;
      if (percent > 1) {
        percent = 1;
      }
      return percent;
    }
    //是否是最后一个可以播放的视频

  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          fragments = _props2.fragments,
          beginDateTime = _props2.beginDateTime,
          selection = _props2.selection,
          isError = _props2.isError,
          end = _props2.end;

      var duration = this.duration;
      // this.setStorage();
      var percent = this.getPercent();
      this.percent = percent;
      return React.createElement(
        Slider,
        {
          ref: 'slider',
          onChange: this.onSliderChange,
          percent: percent,
          padding: '0 10px',
          className: classnames('html5-player-time-slider', 'html5-player-track-history-slider', {
            'html5-player-hide': isError || this.isLastVideo && end
          }),
          onMouseDown: this.onMouseDown,
          onMouseUp: this.onMouseUp,
          onlyShowCircleOnEnter: true
        },
        selection && this.renderSelection(),
        fragments && fragments.map(function (v, k) {
          var gap = v.end - v.begin;
          if (fragments.length === k + 1) {
            //最后一个加多一个1，这样才会覆盖整个slider
            gap += 1;
          }
          var width = gap / duration * 100 + '%';
          var left = v.begin / duration * 100 + '%';
          var style = {
            left: left,
            width: width
          };
          return React.createElement('div', {
            key: k,
            className: classnames({
              'html5-player-broken': !v.file,
              'html5-player-hasvideo': v.file
            }),
            style: style
          });
        }),
        React.createElement(TimeTooltip, { duration: duration, beginDateTime: beginDateTime })
      );
    }
  }, {
    key: 'duration',
    get: function get() {
      var _props3 = this.props,
          duration = _props3.duration,
          sliderModel = _props3.sliderModel;
      var _sliderModel$duration3 = sliderModel.duration,
          currentVideoDuration = _sliderModel$duration3 === undefined ? 0 : _sliderModel$duration3;

      if (currentVideoDuration > duration) {
        //只有一个视频的时候，会存在误差
        duration = currentVideoDuration;
      }
      return duration;
    }
  }, {
    key: 'isLastVideo',
    get: function get() {
      var _props4 = this.props,
          fragments = _props4.fragments,
          activeItem = _props4.activeItem;

      activeItem += 1;
      if (fragments[activeItem] && !fragments[activeItem].file && fragments.length - 1 === activeItem) {
        //判断下一个是否有视频（最后一个视频）
        return true;
      }
      if (fragments && fragments[0] && activeItem < fragments.length) {
        return false;
      }
      return true;
    }
  }]);

  return TimeSlider;
}(React.Component), _class2.propTypes = {}, _class2.displayName = 'TimeSlider', _class2.contextTypes = {
  leftSelectionComponent: PropTypes.element,
  rightSelectionComponent: PropTypes.element
}, _class2.childContextTypes = {
  timeSliderDOM: PropTypes.object
}, _temp2)) || _class) || _class);
export { TimeSlider as default };
