// html5-player/libs/view/controlbar/time-slider/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _dec2, _class, _class2, _temp2; //外部依赖包

//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactRedux = require('react-redux');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _clear = require('../../decorator/clear');

var _clear2 = _interopRequireDefault(_clear);

var _slider = require('../../components/slider');

var _slider2 = _interopRequireDefault(_slider);

var _timeTooltip = require('../../controlbar/time-tooltip');

var _timeTooltip2 = _interopRequireDefault(_timeTooltip);

var _selection = require('./selection');

var _selection2 = _interopRequireDefault(_selection);

var _video = require('../../../model/video');

var _timeSlider = require('../../../model/time-slider');

var _fragment = require('../../../model/fragment');

var _selection3 = require('../../../model/selection');

var _errorMessage = require('../../../model/error-message');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 播放器加载状态的组件
 */
var TimeSlider = (_dec = (0, _reactRedux.connect)(function (state) {
  return {
    sliderModel: state[_timeSlider.namespace],
    fragment: state[_fragment.namespace],
    selection: state[_selection3.namespace],
    isError: !!state[_errorMessage.namespace].message
  };
}), _dec2 = (0, _clear2.default)([_timeSlider.namespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(TimeSlider, _React$Component);

  function TimeSlider() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, TimeSlider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = TimeSlider.__proto__ || (0, _getPrototypeOf2.default)(TimeSlider)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _this.dispatch = _this.props.dispatch, _this.onSliderChange = function (percent) {
      _this.dispatch({
        type: _video.namespace + '/seeking',
        payload: {
          percent: percent
        }
      });
    }, _this.onMouseDown = function (e) {
      _this.dispatch({
        type: _video.namespace + '/seekingState',
        payload: true
      });
    }, _this.onMouseUp = function (e) {
      _this.dispatch({
        type: _video.namespace + '/seekingState',
        payload: false
      });
    }, _this.onLeftSelectionBlur = function (percent) {
      _this.dispatch({
        type: _video.namespace + '/selection',
        payload: {
          type: 'left-blur',
          percent: percent
        }
      });
    }, _this.onLeftSelectionChange = function (percent) {
      _this.dispatch({
        type: _video.namespace + '/selection',
        payload: {
          type: 'left-change',
          percent: percent
        }
      });
    }, _this.onRightSelectionChange = function (percent) {
      _this.dispatch({
        type: _video.namespace + '/selection',
        payload: {
          type: 'right-change',
          percent: percent
        }
      });
    }, _this.onRightSelectionBlur = function (percent) {
      _this.dispatch({
        type: _video.namespace + '/selection',
        payload: {
          type: 'right-blur',
          percent: percent
        }
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(TimeSlider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        timeSliderDOM: this.timeSliderDOM
      };
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.timeSliderDOM = _reactDom2.default.findDOMNode(this.refs.slider);
    }
  }, {
    key: 'renderSelection',

    /**
     * 在有fragment的前提下，渲染可以选择播放的操作按钮
     */
    value: function renderSelection() {
      var _props = this.props,
          _props$selection = _props.selection,
          begin = _props$selection.begin,
          end = _props$selection.end,
          fragment = _props.fragment;

      if (end > fragment.duration) {
        end = fragment.duration;
      }
      var leftPercent = begin / fragment.duration;
      var rightPercent = end / fragment.duration;
      return _react2.default.createElement(
        'span',
        { className: 'html5-player-selection-container' },
        _react2.default.createElement(
          _selection2.default,
          {
            className: 'html5-player-selection-left',
            percent: leftPercent,
            onBlur: this.onLeftSelectionBlur,
            onChange: this.onLeftSelectionChange
          },
          this.context.leftSelectionComponent
        ),
        _react2.default.createElement(
          _selection2.default,
          {
            percent: rightPercent,
            className: 'html5-player-selection-right',
            onChange: this.onRightSelectionChange,
            onBlur: this.onRightSelectionBlur
          },
          this.context.rightSelectionComponent
        )
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          sliderModel = _props2.sliderModel,
          fragment = _props2.fragment,
          timeSliderShowFormat = _props2.timeSliderShowFormat,
          hasFragment = _props2.hasFragment,
          selection = _props2.selection,
          isError = _props2.isError;
      var percent = sliderModel.percent,
          buffer = sliderModel.buffer;

      var bufferStyle = {};
      if (buffer) {
        bufferStyle.width = buffer * 100 + '%';
      }
      if (isError) {
        return false;
      }
      return _react2.default.createElement(
        _slider2.default,
        {
          ref: 'slider',
          onChange: this.onSliderChange,
          percent: fragment.percent || percent,
          padding: '0 10px',
          className: (0, _classnames2.default)('html5-player-time-slider', {
            'html5-player-track-history-slider': hasFragment
          }),
          onMouseDown: this.onMouseDown,
          onMouseUp: this.onMouseUp,
          onlyShowCircleOnEnter: true
        },
        fragment && selection && this.renderSelection(),
        fragment && fragment.data && fragment.data.map(function (v, k) {
          var width = v.gap / fragment.duration * 100 + '%';
          var left = v.begin / fragment.duration * 100 + '%';
          var style = {
            left: left,
            width: width
          };
          return _react2.default.createElement('div', { key: k, className: 'html5-player-broken', style: style });
        }),
        _react2.default.createElement('div', { className: 'html5-player-buffer', style: bufferStyle }),
        _react2.default.createElement(_timeTooltip2.default, { timeSliderShowFormat: timeSliderShowFormat })
      );
    }
  }]);
  return TimeSlider;
}(_react2.default.Component), _class2.propTypes = {}, _class2.displayName = 'TimeSlider', _class2.contextTypes = {
  leftSelectionComponent: _propTypes2.default.element,
  rightSelectionComponent: _propTypes2.default.element
}, _class2.childContextTypes = {
  timeSliderDOM: _propTypes2.default.object
}, _temp2)) || _class) || _class);
exports.default = TimeSlider;
