// html5-player/es/libs/provider/redux-provider.js
import _toConsumableArray from 'babel-runtime/helpers/toConsumableArray';
import _extends from 'babel-runtime/helpers/extends';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _class, _temp2;

import React from 'react';
import PropTypes from 'prop-types';
import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import { Provider } from 'react-redux';
import window from 'global/window';

/**
 * redux 基础 provider
 * @prop { object } store redux store
 *                  非必要的，如果有store，reducers、middlewares和enhancers,preloadedState就不生效，传进来也没有意义
 * @prop { array } middlewares redux middlewares
 * @prop { array } enhancers redux enhancers
 * @prop { object } reducers redux reducers (传进来后会被combineReducers)
 * @prop { any } preloadedState redux preloadedState
 * @prop { boolean } production 是否是生产环境
 */
var ReduxProvider = (_temp2 = _class = function (_React$Component) {
  _inherits(ReduxProvider, _React$Component);

  function ReduxProvider() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ReduxProvider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ReduxProvider.__proto__ || _Object$getPrototypeOf(ReduxProvider)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'Provider', _this.state = {}, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ReduxProvider, [{
    key: 'getStore',
    value: function getStore() {
      var _props = this.props,
          store = _props.store,
          preloadedState = _props.preloadedState;

      if (!store) {
        var reducers = this.getReducers(this.props);
        var enhancers = this.getEnhancers(this.props);
        store = createStore(reducers, preloadedState, enhancers);
      }
      return store;
    }
  }, {
    key: 'getReducers',
    value: function getReducers(props) {
      var reducers = props.reducers;

      return combineReducers(_extends({}, reducers));
    }
  }, {
    key: 'getEnhancers',
    value: function getEnhancers(props) {
      var _props$enhancers = props.enhancers,
          enhancers = _props$enhancers === undefined ? [] : _props$enhancers,
          _props$middlewares = props.middlewares,
          middlewares = _props$middlewares === undefined ? [] : _props$middlewares,
          _props$production = props.production,
          production = _props$production === undefined ? true : _props$production;

      var devtools = function devtools() {
        return function (noop) {
          return noop;
        };
      };
      //如果localStorage.reduxTools=true，reduxTools强制打开。
      //给生产环境调试使用。
      if ((!production || JSON.parse(localStorage.getItem('reduxDevTools'))) && window.__REDUX_DEVTOOLS_EXTENSION__) {
        devtools = window.__REDUX_DEVTOOLS_EXTENSION__;
      } else {
        console.log('You have not install the redux-devtools-extension.');
        console.log('See https://github.com/zalmoxisus/redux-devtools-extension.');
      }
      var _middlewares = [].concat(_toConsumableArray(middlewares));
      var _enhancers = [applyMiddleware.apply(undefined, _toConsumableArray(_middlewares)), devtools()].concat(_toConsumableArray(enhancers));
      return compose.apply(undefined, _toConsumableArray(_enhancers));
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var store = nextProps.store,
          hot = nextProps.hot;
      //热替换处理，根据props.hot来处理

      if (!store && hot) {
        var reducers = this.getReducers(nextProps);
        this.store.replaceReducer(reducers);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.store = null;
    }
  }, {
    key: 'render',
    value: function render() {
      var children = this.props.children;

      if (!this.store) {
        this.store = this.getStore();
      }
      return React.createElement(
        Provider,
        { store: this.store },
        children
      );
    }
  }]);

  return ReduxProvider;
}(React.Component), _class.propTypes = {
  //非必要的，如果有store，reducers、middlewares和enhancers就不生效，传进来也没有意义
  store: PropTypes.object,
  middlewares: PropTypes.array,
  enhancers: PropTypes.array,
  preloadedState: PropTypes.any,
  reducers: PropTypes.object,
  production: PropTypes.bool
}, _temp2);
export { ReduxProvider as default };
