// html5-player/es/view/components/carousel.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _class, _temp2;

//外部依赖包
import React from 'react';
import ReactDOM from 'react-dom';
import classnames from 'classnames';
//import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
//内部依赖包
// import addEventListener from '../../utils/dom/addEventListener';
import { cloneElement, getChildProps } from '../../utils/util';

var defaultGap = 10;
var defaultShowScrollCount = 4;

var Carousel = (_temp2 = _class = function (_React$Component) {
  _inherits(Carousel, _React$Component);

  function Carousel() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Carousel);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Carousel.__proto__ || _Object$getPrototypeOf(Carousel)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'Carousel', _this.state = {
      innerContainerLeft: 0
    }, _this.scrollCount = 0, _this.arrowRightClick = function (e) {
      e.stopPropagation();
      if (_this.scrollCount >= _this.maxScrollCount) {
        //右边最大。
        return;
      }
      _this.scrollCount++;
      _this.setState({
        //触发更新，这里的状态数据，不使用
        innerContainerLeft: _this.innerContainerLeft
      });
    }, _this.arrowLeftClick = function (e) {
      e.stopPropagation();
      if (_this.scrollCount <= 0) {
        //左边到尽头了
        return;
      }
      _this.scrollCount--;
      _this.setState({
        //触发更新，这里的状态数据，不使用
        innerContainerLeft: _this.innerContainerLeft
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Carousel, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var children = this.props.children;

      if (!children) {
        return;
      }
      var containerDOM = ReactDOM.findDOMNode(this.refs['carousel-container']);
      this.containerInnerDOM = ReactDOM.findDOMNode(this.refs['carousel-inner-container']);
      this.carouselWidth = containerDOM.clientWidth;
      this.setState({
        carouselWidth: containerDOM.clientWidth
      });
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var _props = this.props,
          _props$showScrollCoun = _props.showScrollCount,
          showScrollCount = _props$showScrollCoun === undefined ? defaultShowScrollCount : _props$showScrollCoun,
          _props$children = _props.children,
          children = _props$children === undefined ? [] : _props$children;

      if (nextProps.activeItem <= children.length && nextProps.activeItem !== this.props.activeItem) {
        this.scrollCount = Math.ceil(nextProps.activeItem / showScrollCount) - 1;
      }
    }
    //记录向左向右滚动的次数，左减一，右加一。

  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          className = _props2.className,
          _props2$gap = _props2.gap,
          gap = _props2$gap === undefined ? defaultGap : _props2$gap,
          _props2$showScrollCou = _props2.showScrollCount,
          showScrollCount = _props2$showScrollCou === undefined ? defaultShowScrollCount : _props2$showScrollCou,
          _props2$children = _props2.children,
          children = _props2$children === undefined ? [] : _props2$children,
          activeItem = _props2.activeItem;

      if (!children[0]) {
        return false;
      }
      var percent = 1 / showScrollCount;
      var carouselWidth = this.state.carouselWidth;

      var arrowLeftDisable = this.scrollCount === 0;
      var arrowRightDisable = this.scrollCount >= this.maxScrollCount;
      return React.createElement(
        'div',
        { className: 'relative' },
        React.createElement(
          'div',
          {
            className: 'html5-player-carousel-arrow-left',
            onClick: this.arrowLeftClick
          },
          React.createElement(
            'svg',
            {
              className: classnames('html5-player-icon html5-player-arrow-left-icon', {
                'html5-player-icon-disable': arrowLeftDisable
              }),
              'aria-hidden': 'true'
            },
            React.createElement('use', { xlinkHref: '#icon-arrow-left' })
          )
        ),
        React.createElement(
          'div',
          {
            className: classnames('html5-player-carousel', className),
            ref: 'carousel-container'
          },
          carouselWidth && React.createElement(
            'div',
            {
              className: 'html5-player-carousel-inner-contianer html5-player-carousel-transition',
              ref: 'carousel-inner-container',
              style: { left: this.innerContainerLeft }
            },
            children && children.map(function (v, k) {
              var marginLeft = gap + 'px';
              if (k === 0) {
                marginLeft = 0;
              }
              //container宽度 - 间距
              var itemWidth = carouselWidth * percent - gap + gap * percent;
              var childProps = getChildProps(v);
              return cloneElement(v, {
                style: {
                  width: itemWidth + 'px',
                  marginLeft: marginLeft
                },
                className: classnames('html5-player-carousel-item', childProps.className, {
                  'html5-player-carousel-item-active': k + 1 === activeItem
                }),
                key: k
              }, null);
            })
          )
        ),
        React.createElement(
          'div',
          {
            className: 'html5-player-carousel-arrow-right',
            onClick: this.arrowRightClick
          },
          React.createElement(
            'svg',
            {
              className: classnames('html5-player-icon html5-player-arrow-left-icon', {
                'html5-player-icon-disable': arrowRightDisable
              }),
              'aria-hidden': 'true'
            },
            React.createElement('use', { xlinkHref: '#icon-arrow-right' })
          )
        )
      );
    }
  }, {
    key: 'maxScrollCount',
    get: function get() {
      var _props3 = this.props,
          _props3$showScrollCou = _props3.showScrollCount,
          showScrollCount = _props3$showScrollCou === undefined ? defaultShowScrollCount : _props3$showScrollCou,
          _props3$children = _props3.children,
          children = _props3$children === undefined ? [] : _props3$children;

      return Math.ceil(children.length / showScrollCount) - 1;
    }
  }, {
    key: 'innerContainerLeft',
    get: function get() {
      var _props4 = this.props,
          _props4$gap = _props4.gap,
          gap = _props4$gap === undefined ? defaultGap : _props4$gap,
          _props4$children = _props4.children,
          children = _props4$children === undefined ? [] : _props4$children,
          _props4$showScrollCou = _props4.showScrollCount,
          showScrollCount = _props4$showScrollCou === undefined ? defaultShowScrollCount : _props4$showScrollCou;

      var remainder = children.length % showScrollCount;
      var left = -(this.carouselWidth + gap) * this.scrollCount;
      if (this.maxScrollCount !== 0 && this.maxScrollCount === this.scrollCount && remainder !== 0) {
        left = -(this.carouselWidth + gap) * (this.maxScrollCount - 1) - remainder / showScrollCount * this.carouselWidth - gap;
      }
      return left;
    }
  }]);

  return Carousel;
}(React.Component), _class.propTypes = {
  //图片直接的间距
  gap: PropTypes.number,
  //默认可见局域，展示的图片个数
  defaultShowScrollCount: PropTypes.number,
  //当前激活的item
  activeItem: PropTypes.number
}, _temp2);
export { Carousel as default };
