// html5-player/es/view/controlbar/time-slider/index.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _dec, _dec2, _class, _class2, _temp2;

//外部依赖包
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import classnames from 'classnames';
//内部依赖包
import clearDecorator from '../../decorator/clear';
import Slider from '../../components/slider';
import TimeTooltip from '../../controlbar/time-tooltip';
import Selection from './selection';
import { namespace as vidoeNamespace } from '../../../model/video';
import { namespace as timeSliderNamespace } from '../../../model/time-slider';
import { namespace as fragmentNamespace } from '../../../model/fragment';
import { namespace as selectionNamespace } from '../../../model/selection';
import { namespace as errorNamespace } from '../../../model/error-message';

/**
 * 播放器加载状态的组件
 */
var TimeSlider = (_dec = connect(function (state) {
  return {
    sliderModel: state[timeSliderNamespace],
    fragment: state[fragmentNamespace],
    selection: state[selectionNamespace],
    isError: !!state[errorNamespace].message
  };
}), _dec2 = clearDecorator([timeSliderNamespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  _inherits(TimeSlider, _React$Component);

  function TimeSlider() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, TimeSlider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = TimeSlider.__proto__ || _Object$getPrototypeOf(TimeSlider)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _this.dispatch = _this.props.dispatch, _this.onSliderChange = function (percent) {
      _this.dispatch({
        type: vidoeNamespace + '/seeking',
        payload: {
          percent: percent
        }
      });
    }, _this.onMouseDown = function (e) {
      _this.dispatch({
        type: vidoeNamespace + '/seekingState',
        payload: true
      });
    }, _this.onMouseUp = function (e) {
      _this.dispatch({
        type: vidoeNamespace + '/seekingState',
        payload: false
      });
    }, _this.onLeftSelectionBlur = function (percent) {
      _this.dispatch({
        type: vidoeNamespace + '/selection',
        payload: {
          type: 'left-blur',
          percent: percent
        }
      });
    }, _this.onLeftSelectionChange = function (percent) {
      _this.dispatch({
        type: vidoeNamespace + '/selection',
        payload: {
          type: 'left-change',
          percent: percent
        }
      });
    }, _this.onRightSelectionChange = function (percent) {
      _this.dispatch({
        type: vidoeNamespace + '/selection',
        payload: {
          type: 'right-change',
          percent: percent
        }
      });
    }, _this.onRightSelectionBlur = function (percent) {
      _this.dispatch({
        type: vidoeNamespace + '/selection',
        payload: {
          type: 'right-blur',
          percent: percent
        }
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(TimeSlider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        timeSliderDOM: this.timeSliderDOM
      };
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.timeSliderDOM = ReactDOM.findDOMNode(this.refs.slider);
    }
  }, {
    key: 'renderSelection',

    /**
     * 在有fragment的前提下，渲染可以选择播放的操作按钮
     */
    value: function renderSelection() {
      var _props = this.props,
          _props$selection = _props.selection,
          begin = _props$selection.begin,
          end = _props$selection.end,
          fragment = _props.fragment;

      if (end > fragment.duration) {
        end = fragment.duration;
      }
      var leftPercent = begin / fragment.duration;
      var rightPercent = end / fragment.duration;
      return React.createElement(
        'span',
        { className: 'html5-player-selection-container' },
        React.createElement(
          Selection,
          {
            className: 'html5-player-selection-left',
            percent: leftPercent,
            onBlur: this.onLeftSelectionBlur,
            onChange: this.onLeftSelectionChange
          },
          this.context.leftSelectionComponent
        ),
        React.createElement(
          Selection,
          {
            percent: rightPercent,
            className: 'html5-player-selection-right',
            onChange: this.onRightSelectionChange,
            onBlur: this.onRightSelectionBlur
          },
          this.context.rightSelectionComponent
        )
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          sliderModel = _props2.sliderModel,
          fragment = _props2.fragment,
          timeSliderShowFormat = _props2.timeSliderShowFormat,
          hasFragment = _props2.hasFragment,
          selection = _props2.selection,
          isError = _props2.isError;
      var percent = sliderModel.percent,
          buffer = sliderModel.buffer;

      var bufferStyle = {};
      if (buffer) {
        bufferStyle.width = buffer * 100 + '%';
      }
      if (isError) {
        return false;
      }
      return React.createElement(
        Slider,
        {
          ref: 'slider',
          onChange: this.onSliderChange,
          percent: fragment.percent || percent,
          padding: '0 10px',
          className: classnames('html5-player-time-slider', {
            'html5-player-track-history-slider': hasFragment
          }),
          onMouseDown: this.onMouseDown,
          onMouseUp: this.onMouseUp,
          onlyShowCircleOnEnter: true
        },
        fragment && selection && this.renderSelection(),
        fragment && fragment.data && fragment.data.map(function (v, k) {
          var width = v.gap / fragment.duration * 100 + '%';
          var left = v.begin / fragment.duration * 100 + '%';
          var style = {
            left: left,
            width: width
          };
          return React.createElement('div', { key: k, className: 'html5-player-broken', style: style });
        }),
        React.createElement('div', { className: 'html5-player-buffer', style: bufferStyle }),
        React.createElement(TimeTooltip, { timeSliderShowFormat: timeSliderShowFormat })
      );
    }
  }]);

  return TimeSlider;
}(React.Component), _class2.propTypes = {}, _class2.displayName = 'TimeSlider', _class2.contextTypes = {
  leftSelectionComponent: PropTypes.element,
  rightSelectionComponent: PropTypes.element
}, _class2.childContextTypes = {
  timeSliderDOM: PropTypes.object
}, _temp2)) || _class) || _class);
export { TimeSlider as default };
