// html5-player/es/view/controlbar/time-slider/selection.js
import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _class, _temp2;

//外部依赖包
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
//内部依赖包
import addEventListener from '../../../utils/dom/addEventListener';

var Selection = (_temp2 = _class = function (_React$Component) {
  _inherits(Selection, _React$Component);

  function Selection() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Selection);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Selection.__proto__ || _Object$getPrototypeOf(Selection)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _this.dispatch = _this.props.dispatch, _this.events = [], _this.eventsAfterMouseDown = [], _this.onChange = function (percent) {
      var onChange = _this.props.onChange;

      onChange && onChange(percent);
    }, _this.onBlur = function (percent) {
      var onBlur = _this.props.onBlur;

      onBlur && onBlur(percent);
    }, _this.onMouseDown = function (e) {
      e.stopPropagation();
      _this.bindEventsAfterMouseDown();
    }, _this.onMouseUp = function (e) {
      e.stopPropagation();
      _this.removeEventsAfterMouseDown();
      _this.onBlur(_this.percent);
    }, _this.onMouseMove = function (e) {
      _this.onStart(e.pageX);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Selection, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.selectionDOM = ReactDOM.findDOMNode(this.refs['selection']);
      this.document = this.selectionDOM.ownerDocument;
      this.bindEvents();
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var nextPercent = nextProps.percent;
      var thisPercent = this.props.percent;

      if (thisPercent || thisPercent === 0) {
        if (nextPercent !== thisPercent) {
          this.setSliderValueByPercent(nextPercent);
        }
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.removeEvents();
    }
  }, {
    key: 'setSliderValueByPercent',
    value: function setSliderValueByPercent() {
      var percent = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;

      this.percent = percent;
    }
  }, {
    key: 'bindEvents',
    value: function bindEvents() {
      //像mousemove、mousedown、mouseup等事件，直接使用jsx绑定方式，在高德地图上的tooltip会失效。
      this.events.push(addEventListener(this.selectionDOM, 'mousedown', this.onMouseDown));
      this.events.push(addEventListener(this.selectionDOM, 'mouseup', this.onMouseUp));
    }
  }, {
    key: 'removeEvents',
    value: function removeEvents() {
      //移除事件
      this.events.forEach(function (v) {
        v.remove && v.remove();
      });
    }
  }, {
    key: 'bindEventsAfterMouseDown',
    value: function bindEventsAfterMouseDown() {
      var timeSliderDOM = this.context.timeSliderDOM.children[0];
      this.eventsAfterMouseDown.push(addEventListener(this.document, 'mousemove', this.onMouseMove));
      this.eventsAfterMouseDown.push(addEventListener(this.document, 'mouseup', this.onMouseUp));
      this.eventsAfterMouseDown.push(addEventListener(timeSliderDOM, 'onMouseMove', this.onMouseMove));
    }
  }, {
    key: 'removeEventsAfterMouseDown',
    value: function removeEventsAfterMouseDown() {
      //移除事件
      this.eventsAfterMouseDown.forEach(function (v) {
        v.remove && v.remove();
      });
    }
  }, {
    key: 'getSliderStart',
    value: function getSliderStart() {
      var slider = this.context.timeSliderDOM.children[0];
      var rect = slider.getBoundingClientRect();
      return rect.left;
    }
  }, {
    key: 'calcValueByPos',
    value: function calcValueByPos(position) {
      var pixelOffset = position - this.getSliderStart();
      return pixelOffset;
    }
  }, {
    key: 'onStart',
    value: function onStart(position) {
      var value = this.calcValueByPos(position);
      var timeSliderDOM = this.context.timeSliderDOM.children[0];
      var sliderLength = timeSliderDOM.clientWidth;
      if (value < 0) {
        value = 0;
      }
      if (value > sliderLength) {
        value = sliderLength;
      }
      var percent = value / sliderLength;
      var defaultPercent = this.props.defaultPercent;

      if (defaultPercent || defaultPercent === 0) {
        this.percent = percent;
        this.setState({ random: Math.random() });
      }
      this.onChange(percent);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          percent = _props.percent,
          defaultPercent = _props.defaultPercent,
          children = _props.children,
          other = _objectWithoutProperties(_props, ['percent', 'defaultPercent', 'children']);

      if (this.percent === undefined) {
        this.percent = percent || defaultPercent || 0;
      }
      var style = _extends({}, this.props.style);
      style.left = this.percent * 100 + '%';
      return React.createElement(
        'div',
        _extends({}, other, { style: style, ref: 'selection' }),
        children
      );
    }
  }]);

  return Selection;
}(React.Component), _class.propTypes = {}, _class.displayName = 'Selection', _class.contextTypes = {
  timeSliderDOM: PropTypes.object
}, _temp2);
export { Selection as default };
