// html5-player/es/view/end.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _dec, _dec2, _class, _class2, _temp2;

//外部依赖包
import React from 'react';
//import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import classnames from 'classnames';
//内部依赖包
import clearDecorator from './decorator/clear';
import { namespace as endNamespace } from '../model/end';
import { namespace as videoNamespace } from '../model/video';

/**
 * 播放器视频播放结束后的组件
 */
var End = (_dec = connect(function (state) {
  return {
    end: state[endNamespace]
  };
}), _dec2 = clearDecorator([endNamespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  _inherits(End, _React$Component);

  function End() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, End);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = End.__proto__ || _Object$getPrototypeOf(End)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _this.dispatch = _this.props.dispatch, _this.replay = function (e) {
      _this.dispatch({
        type: videoNamespace + '/replay'
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }
  //这里的配置参考jw-player的api


  _createClass(End, [{
    key: 'getClassName',
    value: function getClassName(flag) {
      return classnames('html5-player-cover-view html5-player-end-view', {
        'html5-player-hide': flag
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var end = this.props.end;

      if (!end) {
        return React.createElement('div', { className: this.getClassName(true) });
      }
      return React.createElement(
        'div',
        {
          className: this.getClassName(),
          onDoubleClick: function onDoubleClick(e) {
            e.stopPropagation();
          },
          onClick: function onClick(e) {
            e.stopPropagation();
          }
        },
        React.createElement(
          'button',
          {
            type: 'button',
            className: 'html5-player-middle-button',
            onClick: this.replay
          },
          React.createElement(
            'svg',
            {
              className: 'html5-player-icon html5-player-replay-icon',
              'aria-hidden': 'true'
            },
            React.createElement('use', { xlinkHref: '#icon-replay' })
          )
        )
      );
    }
  }]);

  return End;
}(React.Component), _class2.propTypes = {}, _class2.displayName = 'End', _temp2)) || _class) || _class);
export { End as default };
