// html5-player/es/view/error-message.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _dec, _dec2, _class, _class2, _temp2;

//外部依赖包
import React from 'react';
//import ReactDOM from 'react-dom';
//import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import classnames from 'classnames';
//内部依赖包
import clearDecorator from './decorator/clear';
import { namespace as videoNamespace } from '../model/video';
import { namespace as errorMessageNamespace } from '../model/error-message';

/**
 * 播放器加载状态的组件
 */
var ErrorMessage = (_dec = connect(function (state) {
  return {
    errorInfo: state[errorMessageNamespace]
  };
}), _dec2 = clearDecorator([errorMessageNamespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  _inherits(ErrorMessage, _React$Component);

  function ErrorMessage() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ErrorMessage);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ErrorMessage.__proto__ || _Object$getPrototypeOf(ErrorMessage)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'ErrorMessage', _this.state = {}, _this.dispatch = _this.props.dispatch, _this.reload = function (e) {
      _this.dispatch({
        type: videoNamespace + '/reload'
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }
  //这里的配置参考jw-player的api


  _createClass(ErrorMessage, [{
    key: 'getClassName',
    value: function getClassName(flag) {
      return classnames('html5-player-cover-view html5-player-error-message-view', {
        'html5-player-hide': flag
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var message = this.props.errorInfo.message;

      if (!message) {
        //这里不return false 是为了方便单元测试判断。
        return React.createElement('div', { className: this.getClassName(true) });
      }
      return React.createElement(
        'div',
        {
          className: this.getClassName(),
          onDoubleClick: function onDoubleClick(e) {
            e.stopPropagation();
          },
          onClick: function onClick(e) {
            e.stopPropagation();
          }
        },
        React.createElement(
          'div',
          { className: 'html5-player-error-message-container' },
          React.createElement(
            'button',
            { onClick: this.reload },
            React.createElement(
              'svg',
              {
                className: 'html5-player-icon html5-player-replay-icon html5-player-reload ',
                'aria-hidden': 'true'
              },
              React.createElement('use', { xlinkHref: '#icon-replay' })
            )
          ),
          React.createElement(
            'div',
            null,
            message
          )
        )
      );
    }
  }]);

  return ErrorMessage;
}(React.Component), _class2.propTypes = {}, _temp2)) || _class) || _class);
export { ErrorMessage as default };
