// html5-player/es/view/not-autoplay.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _dec, _dec2, _class, _class2, _temp2;

//外部依赖包
import React from 'react';
//import ReactDOM from 'react-dom';
//import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import classnames from 'classnames';
//内部依赖包
import clearDecorator from './decorator/clear';
import { namespace as notAutoPlayNamespace } from '../model/not-autoplay';
import { namespace as videoNamespace } from '../model/video';

/**
 * 播放器视频播放结束后的组件
 */
var NotAutoPlay = (_dec = connect(function (state) {
  return {
    show: state[notAutoPlayNamespace]
  };
}), _dec2 = clearDecorator([notAutoPlayNamespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  _inherits(NotAutoPlay, _React$Component);

  function NotAutoPlay() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, NotAutoPlay);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = NotAutoPlay.__proto__ || _Object$getPrototypeOf(NotAutoPlay)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'NotAutoPlay', _this.dispatch = _this.props.dispatch, _temp), _possibleConstructorReturn(_this, _ret);
  }
  //这里的配置参考jw-player的api


  _createClass(NotAutoPlay, [{
    key: 'getClassName',
    value: function getClassName(flag) {
      return classnames('html5-player-cover-view html5-player-play-view', {
        'html5-player-hide': flag
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var show = this.props.show;

      if (!show) {
        //这里不return false 是为了方便单元测试判断。
        return React.createElement('div', { className: this.getClassName(true) });
      }
      return React.createElement(
        'div',
        {
          className: this.getClassName(),
          onDoubleClick: function onDoubleClick(e) {
            e.stopPropagation();
          },
          onClick: function onClick(e) {
            e.stopPropagation();
          }
        },
        show && React.createElement(
          'button',
          {
            className: 'html5-player-middle-button',
            onClick: function onClick(e) {
              e.stopPropagation();
              _this2.dispatch({
                type: videoNamespace + '/playAfterNotAutoplay'
              });
            }
          },
          React.createElement(
            'svg',
            {
              className: 'html5-player-icon html5-player-play-big-icon',
              'aria-hidden': 'true'
            },
            React.createElement('use', { xlinkHref: '#icon-play' })
          )
        )
      );
    }
  }]);

  return NotAutoPlay;
}(React.Component), _class2.propTypes = {}, _temp2)) || _class) || _class);
export { NotAutoPlay as default };
