// html5-player/libs/view/controlbar/next.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp2; //外部依赖包

//import ReactDOM from 'react-dom';


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

//内部依赖包

var Next = (_temp2 = _class = function (_React$Component) {
  (0, _inherits3.default)(Next, _React$Component);

  function Next() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Next);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Next.__proto__ || (0, _getPrototypeOf2.default)(Next)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'Next', _this.onClick = function (e) {
      _this.context.setActiveItem(_this.context.activeItem + 1);
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Next, [{
    key: 'render',
    value: function render() {
      var _context = this.context,
          activeItem = _context.activeItem,
          playlist = _context.playlist;

      if (!playlist || !playlist[0]) {
        return false;
      }
      if (activeItem >= playlist.length) {
        return false;
      }
      return _react2.default.createElement(
        'button',
        {
          type: 'button',
          className: 'html5-player-small-button float-left html5-player-next-button',
          onClick: this.onClick
        },
        _react2.default.createElement(
          'svg',
          {
            className: (0, _classnames2.default)('html5-player-icon html5-player-next-icon'),
            'aria-hidden': 'true'
          },
          _react2.default.createElement('use', { xlinkHref: '#icon-next' })
        )
      );
    }
  }]);
  return Next;
}(_react2.default.Component), _class.contextTypes = {
  playlist: _propTypes2.default.array,
  activeItem: _propTypes2.default.number,
  setActiveItem: _propTypes2.default.func
}, _temp2);
exports.default = Next;
