// html5-player/libs/view/controlbar/play-pause.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _dec2, _class, _class2, _temp2; //外部依赖包

//import ReactDOM from 'react-dom';
//import PropTypes from 'prop-types';

//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _clear = require('../decorator/clear');

var _clear2 = _interopRequireDefault(_clear);

var _video = require('../../model/video');

var _playPause = require('../../model/play-pause');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 播放器加载状态的组件
 */
var PlayPause = (_dec = (0, _reactRedux.connect)(function (state) {
  return {
    playing: state[_playPause.namespace]
  };
}), _dec2 = (0, _clear2.default)([_playPause.namespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(PlayPause, _React$Component);

  function PlayPause() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, PlayPause);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = PlayPause.__proto__ || (0, _getPrototypeOf2.default)(PlayPause)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'PlayPause', _this.state = {}, _this.dispatch = _this.props.dispatch, _this.play = function (e) {
      e.stopPropagation();
      _this.dispatch({
        type: _video.namespace + '/play',
        payload: {
          noControlbarAction: true
        }
      });
    }, _this.pause = function (e) {
      e.stopPropagation();
      _this.dispatch({
        type: _video.namespace + '/pause'
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(PlayPause, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          playing = _props.playing,
          living = _props.living;

      if (living) {
        return false;
      }
      return _react2.default.createElement(
        'button',
        {
          type: 'button',
          className: 'html5-player-small-button html5-player-play-pause-button',
          onClick: playing ? this.pause : this.play
        },
        _react2.default.createElement(
          'svg',
          {
            className: (0, _classnames2.default)('html5-player-icon html5-player-play-pause-icon', {
              'html5-player-pause-icon': playing,
              'html5-player-play-icon': !playing
            }),
            'aria-hidden': 'true'
          },
          _react2.default.createElement('use', {
            xlinkHref: (0, _classnames2.default)({
              '#icon-pause': playing,
              '#icon-play': !playing
            })
          })
        )
      );
    }
  }]);
  return PlayPause;
}(_react2.default.Component), _class2.propTypes = {}, _temp2)) || _class) || _class);
exports.default = PlayPause;
