// html5-player/es/view/history/time-slider/index.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _dec, _dec2, _class, _class2, _temp2;

//外部依赖包
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import classnames from 'classnames';
//内部依赖包
import clearDecorator from '../../decorator/clear';
import Slider from '../../components/slider';
import TimeTooltip from './time-tooltip';
import Selection from './selection';
import { namespace as videoNamespace } from '../../../model/video';
import { namespace as timeSliderNamespace } from '../../../model/time-slider';
import { namespace as selectionNamespace } from '../../../model/selection';
import { namespace as errorNamespace } from '../../../model/error-message';
import { namespace as historyNamespace } from '../../../model/history';

/**
 * 播放器加载状态的组件
 */
var TimeSlider = (_dec = connect(function (state) {
  return {
    sliderModel: state[timeSliderNamespace],
    selection: state[selectionNamespace],
    isError: !!state[errorNamespace].message,
    historyState: state[historyNamespace]
  };
}), _dec2 = clearDecorator([timeSliderNamespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  _inherits(TimeSlider, _React$Component);

  function TimeSlider() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, TimeSlider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = TimeSlider.__proto__ || _Object$getPrototypeOf(TimeSlider)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _this.dispatch = _this.props.dispatch, _this.onSliderChange = function (percent) {
      if (_this.isSeekingDisabled) {
        return;
      }
      _this.dispatch({
        type: historyNamespace + '/setSliderPercent',
        payload: {
          percent: percent
        }
      });
    }, _this.onMouseDown = function (e) {
      if (_this.isSeekingDisabled) {
        return;
      }
      _this.dispatch({
        type: videoNamespace + '/seekingState',
        payload: true
      });
    }, _this.onMouseUp = function (e) {
      if (_this.isSeekingDisabled) {
        return;
      }
      _this.dispatch({
        type: videoNamespace + '/seekingState',
        payload: false
      });
    }, _this.onLeftSelectionBlur = function (percent) {
      _this.dispatch({
        type: videoNamespace + '/selection',
        payload: {
          type: 'left-blur',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _this.onLeftSelectionChange = function (percent) {
      _this.dispatch({
        type: videoNamespace + '/selection',
        payload: {
          type: 'left-change',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _this.onRightSelectionChange = function (percent) {
      _this.dispatch({
        type: videoNamespace + '/selection',
        payload: {
          type: 'right-change',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _this.onRightSelectionBlur = function (percent) {
      _this.dispatch({
        type: videoNamespace + '/selection',
        payload: {
          type: 'right-blur',
          percent: percent,
          duration: _this.duration
        }
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(TimeSlider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        timeSliderDOM: this.timeSliderDOM
      };
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.timeSliderDOM = ReactDOM.findDOMNode(this.refs.slider);
    }
  }, {
    key: 'renderSelection',

    /**
     * 在有fragment的前提下，渲染可以选择播放的操作按钮
     */
    value: function renderSelection() {
      var selection = this.props.selection;

      if (!selection) {
        return false;
      }
      var begin = selection.begin,
          end = selection.end;

      var duration = this.duration;
      if (end > duration) {
        end = duration;
      }
      var leftPercent = begin / duration;
      var rightPercent = end / duration;
      return React.createElement(
        'span',
        { className: 'html5-player-selection-container' },
        React.createElement(
          Selection,
          {
            className: 'html5-player-selection-left',
            percent: leftPercent,
            onBlur: this.onLeftSelectionBlur,
            onChange: this.onLeftSelectionChange
          },
          this.context.leftSelectionComponent
        ),
        React.createElement(
          Selection,
          {
            percent: rightPercent,
            className: 'html5-player-selection-right',
            onChange: this.onRightSelectionChange,
            onBlur: this.onRightSelectionBlur
          },
          this.context.rightSelectionComponent
        )
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          fragments = _props.fragments,
          beginDateTime = _props.beginDateTime,
          selection = _props.selection,
          isError = _props.isError,
          percent = _props.historyState.percent;

      var duration = this.duration;
      return React.createElement(
        Slider,
        {
          ref: 'slider',
          onChange: this.onSliderChange,
          percent: percent,
          padding: '0 10px',
          className: classnames('html5-player-time-slider', 'html5-player-track-history-slider', {
            'html5-player-hide': isError
          }),
          onMouseDown: this.onMouseDown,
          onMouseUp: this.onMouseUp,
          onlyShowCircleOnEnter: true
        },
        selection && this.renderSelection(),
        React.createElement(
          'div',
          {
            style: {
              overflow: 'hidden',
              position: 'absolute',
              height: '100%',
              width: '100%'
            }
          },
          fragments && fragments.map(function (v, k) {
            var gap = v.end - v.begin;

            var width = gap / duration * 100 + '%';
            if (fragments.length === k + 1) {
              //这样才会覆盖整个slider
              width = (gap / duration + 0.1) * 100 + '%';
            }
            var left = v.begin / duration * 100 + '%';
            var style = {
              left: left,
              width: width
            };
            return React.createElement('div', {
              key: k,
              className: classnames({
                'html5-player-broken': !v.file,
                'html5-player-hasvideo': v.file
              }),
              style: style
            });
          })
        ),
        React.createElement(TimeTooltip, { duration: duration, beginDateTime: beginDateTime })
      );
    }
  }, {
    key: 'isSeekingDisabled',
    get: function get() {
      var selection = this.props.selection;

      if (selection && selection.seekingDisabled) {
        return true;
      }
      return false;
    }
  }, {
    key: 'duration',
    get: function get() {
      var _props2 = this.props,
          duration = _props2.duration,
          sliderModel = _props2.sliderModel;
      var _sliderModel$duration = sliderModel.duration,
          currentVideoDuration = _sliderModel$duration === undefined ? 0 : _sliderModel$duration;

      if (currentVideoDuration > duration) {
        //只有一个视频的时候，会存在误差
        duration = currentVideoDuration;
      }
      return duration;
    }
  }]);

  return TimeSlider;
}(React.Component), _class2.propTypes = {}, _class2.displayName = 'TimeSlider', _class2.contextTypes = {
  leftSelectionComponent: PropTypes.element,
  rightSelectionComponent: PropTypes.element
}, _class2.childContextTypes = {
  timeSliderDOM: PropTypes.object
}, _temp2)) || _class) || _class);
export { TimeSlider as default };
