// html5-player/libs/view/error-message.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _dec2, _class, _class2, _temp2; //外部依赖包

//import ReactDOM from 'react-dom';
//import PropTypes from 'prop-types';

//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _clear = require('./decorator/clear');

var _clear2 = _interopRequireDefault(_clear);

var _video = require('../model/video');

var _errorMessage = require('../model/error-message');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 播放器加载状态的组件
 */
var ErrorMessage = (_dec = (0, _reactRedux.connect)(function (state) {
  return {
    errorInfo: state[_errorMessage.namespace]
  };
}), _dec2 = (0, _clear2.default)([_errorMessage.namespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(ErrorMessage, _React$Component);

  function ErrorMessage() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ErrorMessage);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ErrorMessage.__proto__ || (0, _getPrototypeOf2.default)(ErrorMessage)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'ErrorMessage', _this.state = {}, _this.dispatch = _this.props.dispatch, _this.reload = function (e) {
      _this.dispatch({
        type: _video.namespace + '/reload'
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }
  //这里的配置参考jw-player的api


  (0, _createClass3.default)(ErrorMessage, [{
    key: 'getClassName',
    value: function getClassName(flag) {
      return (0, _classnames2.default)('html5-player-cover-view html5-player-error-message-view', {
        'html5-player-hide': flag
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var message = this.props.errorInfo.message;

      if (!message) {
        //这里不return false 是为了方便单元测试判断。
        return _react2.default.createElement('div', { className: this.getClassName(true) });
      }
      return _react2.default.createElement(
        'div',
        {
          className: this.getClassName(),
          onDoubleClick: function onDoubleClick(e) {
            e.stopPropagation();
          },
          onClick: function onClick(e) {
            e.stopPropagation();
          }
        },
        _react2.default.createElement(
          'div',
          { className: 'html5-player-error-message-container' },
          _react2.default.createElement(
            'button',
            { onClick: this.reload },
            _react2.default.createElement(
              'svg',
              {
                className: 'html5-player-icon html5-player-replay-icon html5-player-reload ',
                'aria-hidden': 'true'
              },
              _react2.default.createElement('use', { xlinkHref: '#icon-replay' })
            )
          ),
          _react2.default.createElement(
            'div',
            null,
            message
          )
        )
      );
    }
  }]);
  return ErrorMessage;
}(_react2.default.Component), _class2.propTypes = {}, _temp2)) || _class) || _class);
exports.default = ErrorMessage;
