// html5-player/libs/view/not-autoplay.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _dec2, _class, _class2, _temp2; //外部依赖包

//import ReactDOM from 'react-dom';
//import PropTypes from 'prop-types';

//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _clear = require('./decorator/clear');

var _clear2 = _interopRequireDefault(_clear);

var _notAutoplay = require('../model/not-autoplay');

var _video = require('../model/video');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 播放器视频播放结束后的组件
 */
var NotAutoPlay = (_dec = (0, _reactRedux.connect)(function (state) {
  return {
    show: state[_notAutoplay.namespace]
  };
}), _dec2 = (0, _clear2.default)([_notAutoplay.namespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(NotAutoPlay, _React$Component);

  function NotAutoPlay() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, NotAutoPlay);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = NotAutoPlay.__proto__ || (0, _getPrototypeOf2.default)(NotAutoPlay)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'NotAutoPlay', _this.dispatch = _this.props.dispatch, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }
  //这里的配置参考jw-player的api


  (0, _createClass3.default)(NotAutoPlay, [{
    key: 'getClassName',
    value: function getClassName(flag) {
      return (0, _classnames2.default)('html5-player-cover-view html5-player-play-view', {
        'html5-player-hide': flag
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var show = this.props.show;

      if (!show) {
        //这里不return false 是为了方便单元测试判断。
        return _react2.default.createElement('div', { className: this.getClassName(true) });
      }
      return _react2.default.createElement(
        'div',
        {
          className: this.getClassName(),
          onDoubleClick: function onDoubleClick(e) {
            e.stopPropagation();
          },
          onClick: function onClick(e) {
            e.stopPropagation();
          }
        },
        show && _react2.default.createElement(
          'button',
          {
            className: 'html5-player-middle-button',
            onClick: function onClick(e) {
              e.stopPropagation();
              _this2.dispatch({
                type: _video.namespace + '/playAfterNotAutoplay'
              });
            }
          },
          _react2.default.createElement(
            'svg',
            {
              className: 'html5-player-icon html5-player-play-big-icon',
              'aria-hidden': 'true'
            },
            _react2.default.createElement('use', { xlinkHref: '#icon-play' })
          )
        )
      );
    }
  }]);
  return NotAutoPlay;
}(_react2.default.Component), _class2.propTypes = {}, _temp2)) || _class) || _class);
exports.default = NotAutoPlay;
