// html5-player/libs/utils/logger/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.warn = exports.error = exports.info = exports.success = exports.log = undefined;
exports.default = logger;

var _const = require('../const');

function logger(type) {
  var originType = type;
  function noop() {}
  if (!_const.DEBUG) {
    return noop;
  }
  if (!window.console) {
    return noop;
  }
  if (!window.console[type]) {
    if (window.console.log) {
      type = 'log';
    } else {
      return noop;
    }
  }
  function log(title) {
    var _window$console;

    var prevArg = [];
    if (Object.prototype.toString.apply(title) === '[object String]' || Object.prototype.toString.apply(title) === '[object Number]') {
      var fontSize = 'font-size: 14px;';
      prevArg.push('%c' + title + '%c');
      if (originType === 'info') {
        prevArg.push(fontSize + 'font-weight: bold;color: #108ee9;');
      } else if (originType === 'success') {
        prevArg.push(fontSize + 'font-weight: bold;color: green;');
      } else {
        prevArg.push(fontSize + 'font-weight: bold');
      }
      prevArg.push('');
    } else {
      prevArg.push(title);
    }

    for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    (_window$console = window.console)[type].apply(_window$console, prevArg.concat(args));
  }
  return log;
}

var log = exports.log = logger('log');
var success = exports.success = logger('success');
var info = exports.info = logger('info');
var error = exports.error = logger('error');
var warn = exports.warn = logger('warn');
