// html5-player/es/view/controlbar/next.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _class, _temp2;

//外部依赖包
import React from 'react';
//import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classnames from 'classnames';
//内部依赖包

var Next = (_temp2 = _class = function (_React$Component) {
  _inherits(Next, _React$Component);

  function Next() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Next);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Next.__proto__ || _Object$getPrototypeOf(Next)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'Next', _this.onClick = function (e) {
      _this.context.setActiveItem(_this.context.activeItem + 1);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Next, [{
    key: 'render',
    value: function render() {
      var _context = this.context,
          activeItem = _context.activeItem,
          playlist = _context.playlist;

      if (!playlist || !playlist[0]) {
        return false;
      }
      if (activeItem >= playlist.length) {
        return false;
      }
      return React.createElement(
        'button',
        {
          type: 'button',
          className: 'html5-player-small-button float-left html5-player-next-button',
          onClick: this.onClick
        },
        React.createElement(
          'svg',
          {
            className: classnames('html5-player-icon html5-player-next-icon'),
            'aria-hidden': 'true'
          },
          React.createElement('use', { xlinkHref: '#icon-next' })
        )
      );
    }
  }]);

  return Next;
}(React.Component), _class.contextTypes = {
  playlist: PropTypes.array,
  activeItem: PropTypes.number,
  setActiveItem: PropTypes.func
}, _temp2);
export { Next as default };
