// html5-player/es/view/controlbar/picture-quality.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _dec, _dec2, _class;

//外部依赖包
import React from 'react';
//import ReactDOM from 'react-dom';
//import PropTypes from 'prop-types';
import { connect } from 'react-redux';
// import classnames from 'classnames';
//内部依赖包
import Tooltip from '../components/tooltip';
import clearDecorator from '../decorator/clear';
import PictureQualityList from './setting/picture-quality-list';
import { namespace as qualityNamespace } from '../../model/picture-quality';

var PictureQuality = (_dec = connect(function (state) {
  return {
    qualityList: state[qualityNamespace].list
  };
}), _dec2 = clearDecorator([qualityNamespace]), _dec(_class = _dec2(_class = function (_React$Component) {
  _inherits(PictureQuality, _React$Component);

  function PictureQuality() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, PictureQuality);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = PictureQuality.__proto__ || _Object$getPrototypeOf(PictureQuality)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'PictureQuality', _this.state = {}, _this.onSelect = function (value) {
      _this.setState({
        tooltipKey: value
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(PictureQuality, [{
    key: 'renderContent',
    value: function renderContent() {
      return React.createElement(PictureQualityList, { onSelect: this.onSelect });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          qualityList = _props.qualityList,
          locale = _props.locale;
      var tooltipKey = this.state.tooltipKey;

      if (!qualityList) {
        return false;
      }
      return React.createElement(
        Tooltip,
        {
          key: tooltipKey,
          trigger: 'click',
          content: this.renderContent(),
          toTargetGap: 20
        },
        React.createElement(
          'span',
          { className: 'float-right cursor-pointer html5-player-common-button html5-player-quality-button' },
          locale.pictureQuality
        )
      );
    }
  }]);

  return PictureQuality;
}(React.Component)) || _class) || _class);
export { PictureQuality as default };
