// html5-player/libs/view/track/subtitle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _dec2, _class, _class2, _temp2; //外部依赖包

//import ReactDOM from 'react-dom';
//import PropTypes from 'prop-types';

//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _clear = require('../decorator/clear');

var _clear2 = _interopRequireDefault(_clear);

var _track = require('../../model/track');

var _time = require('../../model/time');

var _end = require('../../model/end');

var _errorMessage = require('../../model/error-message');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 播放器加载状态的组件
 */
var Subtitle = (_dec = (0, _reactRedux.connect)(function (state) {
  return {
    track: state[_track.namespace],
    time: state[_time.namespace],
    end: state[_end.namespace],
    isError: !!state[_errorMessage.namespace].message
  };
}), _dec2 = (0, _clear2.default)([_track.namespace]), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_React$Component) {
  (0, _inherits3.default)(Subtitle, _React$Component);

  function Subtitle() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Subtitle);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Subtitle.__proto__ || (0, _getPrototypeOf2.default)(Subtitle)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'Subtitle', _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Subtitle, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          track = _props.track,
          time = _props.time,
          userActive = _props.userActive,
          end = _props.end,
          isError = _props.isError;
      var subtitleCues = track.subtitleCues;

      var currentTime = time.currentTime;
      if (!subtitleCues || isError) {
        return _react2.default.createElement('span', { className: 'html5-player-subtitle-text html5-player-hide' });
      }
      var text = void 0;
      subtitleCues.forEach(function (v) {
        if (currentTime >= v.begin && currentTime < v.end) {
          text = v.text;
        }
      });
      //console.log(text,currentTime)
      if (!text || end) {
        //text为空或者undefined
        return false;
      }
      //回车键需要处理成<br/>
      var textArray = text.split('\r\n');
      if (textArray.length === 1) {
        textArray = text.split('\n');
      }
      return _react2.default.createElement(
        'span',
        {
          className: (0, _classnames2.default)('html5-player-subtitle-text', {
            'html5-player-subtitle-text-user-active': userActive,
            'html5-player-subtitle-text-user-inactive': !userActive
          })
        },
        textArray.map(function (v, k) {
          return _react2.default.createElement(
            'span',
            { key: k },
            '\xA0',
            v,
            '\xA0',
            _react2.default.createElement('br', null)
          );
        })
      );
    }
  }]);
  return Subtitle;
}(_react2.default.Component), _class2.propTypes = {}, _temp2)) || _class) || _class);
exports.default = Subtitle;
