// html5-player/libs/view/controlbar/picture-quality.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _dec2, _class; //外部依赖包

//import ReactDOM from 'react-dom';
//import PropTypes from 'prop-types';

// import classnames from 'classnames';
//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _tooltip = require('../components/tooltip');

var _tooltip2 = _interopRequireDefault(_tooltip);

var _clear = require('../decorator/clear');

var _clear2 = _interopRequireDefault(_clear);

var _pictureQualityList = require('./setting/picture-quality-list');

var _pictureQualityList2 = _interopRequireDefault(_pictureQualityList);

var _pictureQuality = require('../../model/picture-quality');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PictureQuality = (_dec = (0, _reactRedux.connect)(function (state) {
  return {
    qualityList: state[_pictureQuality.namespace].list
  };
}), _dec2 = (0, _clear2.default)([_pictureQuality.namespace]), _dec(_class = _dec2(_class = function (_React$Component) {
  (0, _inherits3.default)(PictureQuality, _React$Component);

  function PictureQuality() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, PictureQuality);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = PictureQuality.__proto__ || (0, _getPrototypeOf2.default)(PictureQuality)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'PictureQuality', _this.state = {}, _this.onSelect = function (value) {
      _this.setState({
        tooltipKey: value
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(PictureQuality, [{
    key: 'renderContent',
    value: function renderContent() {
      return _react2.default.createElement(_pictureQualityList2.default, { onSelect: this.onSelect });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          qualityList = _props.qualityList,
          locale = _props.locale;
      var tooltipKey = this.state.tooltipKey;

      if (!qualityList) {
        return false;
      }
      return _react2.default.createElement(
        _tooltip2.default,
        {
          key: tooltipKey,
          trigger: 'click',
          content: this.renderContent(),
          toTargetGap: 20
        },
        _react2.default.createElement(
          'span',
          { className: 'float-right cursor-pointer html5-player-common-button html5-player-quality-button' },
          locale.pictureQuality
        )
      );
    }
  }]);
  return PictureQuality;
}(_react2.default.Component)) || _class) || _class);
exports.default = PictureQuality;
