// html5-player/libs/view/controlbar/playback-rate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _dec2, _class; //外部依赖包

//import ReactDOM from 'react-dom';
//import PropTypes from 'prop-types';

// import classnames from 'classnames';
//内部依赖包


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _tooltip = require('../components/tooltip');

var _tooltip2 = _interopRequireDefault(_tooltip);

var _clear = require('../decorator/clear');

var _clear2 = _interopRequireDefault(_clear);

var _playbackRateList = require('./setting/playback-rate-list');

var _playbackRateList2 = _interopRequireDefault(_playbackRateList);

var _playbackRate = require('../../model/playback-rate');

var _living = require('../../model/living');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PlaybackRate = (_dec = (0, _reactRedux.connect)(function (state) {
  return {
    playbackRate: state[_playbackRate.namespace],
    living: state[_living.namespace]
  };
}), _dec2 = (0, _clear2.default)([_playbackRate.namespace]), _dec(_class = _dec2(_class = function (_React$Component) {
  (0, _inherits3.default)(PlaybackRate, _React$Component);

  function PlaybackRate() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, PlaybackRate);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = PlaybackRate.__proto__ || (0, _getPrototypeOf2.default)(PlaybackRate)).call.apply(_ref, [this].concat(args))), _this), _this.displayName = 'PlaybackRate', _this.state = {}, _this.onRatteSelect = function (rate) {
      _this.setState({
        tooltipKey: rate
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(PlaybackRate, [{
    key: 'renderContent',
    value: function renderContent() {
      return _react2.default.createElement(_playbackRateList2.default, (0, _extends3.default)({}, this.props, { onSelect: this.onRatteSelect }));
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          playbackRate = _props.playbackRate,
          locale = _props.locale;
      var tooltipKey = this.state.tooltipKey;

      return _react2.default.createElement(
        _tooltip2.default,
        {
          key: tooltipKey,
          trigger: 'click',
          content: this.renderContent(),
          toTargetGap: 20
        },
        _react2.default.createElement(
          'span',
          { className: 'float-right cursor-pointer html5-player-rate-button' },
          playbackRate + locale.speed
        )
      );
    }
  }]);
  return PlaybackRate;
}(_react2.default.Component)) || _class) || _class);
exports.default = PlaybackRate;
